src/nameservice/ClickResolver.sol

```diff
@@ -0,0 +1,262 @@
+    }
+
+    /// @notice Calculates storage slot of the key in the L2 registry
+    /// @dev Names are stored in the L2 registry, in a mapping with slot `mappingSlot`
+    function getDomainSlot(bytes32 _key) public view returns (bytes32) {
+        return keccak256(abi.encode(_key, mappingSlot));
+    }
+
+    /// @notice Helper function to convert a string to lowercase
+    function toLower(string memory str) private pure returns (string memory) {
+        bytes memory bStr = bytes(str);
+        bytes memory bLower = new bytes(bStr.length);
+        for (uint256 i = 0; i < bStr.length; i++) {
+            // Uppercase character...
+            if ((uint8(bStr[i]) >= 65) && (uint8(bStr[i]) <= 90)) {
+                // So we add 32 to make it lowercase
+                bLower[i] = bytes1(uint8(bStr[i]) + 32);
+            } else {
+                bLower[i] = bStr[i];
+            }
+        }
+        return string(bLower);
+    }
+
+    /// @notice Calculates storage slot of the key in the L2 registry
+    /// @dev Names are stored in the L2 registry, in a mapping with slot `mappingSlot`
+    function getDomainSlot(string memory _key) public view returns (bytes32) {
+        string memory domain = toLower(_key);
+        uint256 tokenId = uint256(keccak256(abi.encodePacked(domain)));
+        return keccak256(abi.encode(tokenId, mappingSlot));
+    }
+
+    /// @notice Resolves a name to a value
+    /// @param _name The name to resolve, DNS encoded
+    /// @param _data The ABI encoded data for the underlying resolution function (Eg, addr(bytes32), text(bytes32,string), etc).
+    function resolve(bytes calldata _name, bytes calldata _data) external view returns (bytes memory) {
+        bytes memory callData = abi.encodeWithSelector(IOffchainResolver.resolve.selector, _name, _data);
+
+        // Fill URLs
+        string[] memory urls = new string[](1);
+        urls[0] = url;
+
+        (string memory sub, string memory dom, string memory top) = parseDnsDomain(_name);
+
+        // If there's no domain or top level domain, throw
+        if (bytes(dom).length == 0 || bytes(top).length == 0) {
+            revert InvalidDnsDomain();
+        }
+
+        if (bytes(sub).length == 0) {
+            // If there's no subdomain, return the domain owner
+            return abi.encodePacked(domainOwner);
+        }
+
+        bytes32 registrySlot = getDomainSlot(sub);
+
+        bytes4 functionSelector = bytes4(_data[:4]);
+        if (functionSelector == ADDR_SELECTOR) {
+            revert OffchainLookup(
+                address(this),
+                urls,
+                callData,
+                ClickResolver.resolveWithProof.selector,
+                abi.encode(registry, registrySlot)
+            );
+        } else if (functionSelector == ADDR_MULTICHAIN_SELECTOR) {
+            (, uint256 coinType) = abi.decode(_data[4:], (bytes32, uint256));
+            if (coinType != ZKSYNC_MAINNET_COIN_TYPE) {
```

Comment:
(discussed today, adding here for tracking) Shouldn't we have a coin type for testnet too?

Reply:
This ADDR_MULTICHAIN_SELECTOR path is not used for us now. So no need to do anything for testnet at the moment.